src/core/Events.js

```diff
@@ -90,95 +90,116 @@ L.Evented = L.Class.extend({
+			typeListeners.fnIndices[contextId] = fnIndices;
+		}
 
-			events[type] = events[type] || [];
-			events[type].push({fn: fn});
+		// check if fn already there
+		var fnStamp = L.stamp(fn);
+		if (fnIndices[fnStamp]) {
+			return;
 		}
+
+		listeners.push(newListener);
+		fnIndices[fnStamp] = listeners.length - 1;
+		typeListeners.count++;
 	},
 
 	_off: function (type, fn, context) {
-		var events = this._events,
-		    indexKey = type + '_idx',
-		    indexLenKey = type + '_len',
-		    listener, listeners, i, len;
+		var typeListeners,
+		    contextId,
+		    listeners,
+		    i,
+		    len;
 
-		if (!events) { return; }
+		if (!this._events) { return; }
 
 		if (!fn) {
-			// clear all listeners for a type if function isn't specified
-			// set the removed listeners to noop so that's not called if remove happens in fire
-			listeners = events[indexKey];
-			for (i in listeners) {
-				listeners[i].fn = L.Util.falseFn;
-			}
-			listeners = events[type] || [];
-			for (i = 0, len = listeners.length; i < len; i++) {
-				listeners[i].fn = L.Util.falseFn;
+			// Set all removed listeners to noop so they are not called if remove happens in fire
+			typeListeners = this._events[type];
+			if (typeListeners) {
+				for (contextId in typeListeners.listeners) {
+					listeners = typeListeners.listeners[contextId];
+					for (i = 0, len = listeners.length; i < len; i++) {
+						listeners[i].fn = L.Util.falseFn;
+					}
+				}
+				// clear all listeners for a type if function isn't specified
+				delete this._events[type];
 			}
-
-			delete events[type];
-			delete events[indexKey];
-			delete events[indexLenKey];
 			return;
 		}
 
-		var contextId = context && context !== this && L.stamp(context),
-		    id;
-
-		if (contextId) {
-			id = L.stamp(fn) + '_' + contextId;
-			listeners = events[indexKey];
+		typeListeners = this._events[type];
+		if (!typeListeners) {
+			return;
+		}
 
-			if (listeners && listeners[id]) {
-				listener = listeners[id];
-				delete listeners[id];
-				events[indexLenKey]--;
-			}
+		contextId = context && context !== this && L.stamp(context);
+		if (!contextId) {
+			contextId = 'no_context';
+		}
 
-		} else {
-			listeners = events[type];
-
-			if (listeners) {
-				for (i = 0, len = listeners.length; i < len; i++) {
-					if (listeners[i].fn === fn) {
-						listener = listeners[i];
-						listeners.splice(i, 1);
-						break;
+		listeners = typeListeners.listeners[contextId];
+		if (listeners) {
+			var fnStamp = L.stamp(fn),
+			    fnIndices = typeListeners.fnIndices[contextId],
+			    listenerIndex = fnIndices[fnStamp];
+
+			if (listenerIndex >= 0) {
+				// find fn and remove it
+				var l = listeners[listenerIndex];
+				// set the removed listener to noop so that's not called if remove happens in fire
+				l.fn = L.Util.falseFn;
+				typeListeners.count--;
+
+				if (listeners.length > 1) {
+					if (!this._isFiring) {
+						listeners.splice(listenerIndex, 1);

```

Comment:
Humm, I see a sort of issue in altering the `listeners` array while relying on indices to retrieve the listeners.